### ISSUE TITLE
[RFC]: Compact candidate-set renormalization for native top-k/top-p sampling

### ISSUE THREAD

**Issue by Kiid-A** [vllm-project/vllm#46738]
## Motivation

  I would like to discuss a native sampling optimization for the `top_k` / `top_p` path.

  In the current native sampler, the common flow is:

  1. apply top-k / top-p filtering over the full vocabulary;
  2. compute full-vocab softmax;
  3. sample from the full-vocab probability tensor.

  For top-k and top-k+top-p decoding, most vocabulary entries are guaranteed to have zero probability after filtering. That means the later renormalization and sampling work can be done directly on
  the surviving candidate set instead of the full vocabulary.

  The proposed optimization is:

  1. select the surviving candidates;
  2. renormalize only those candidates;
  3. sample from the compact candidate distribution;
  4. map the sampled candidate back to the original vocab id.

  This avoids full-vocab masking + full-vocab softmax + full-vocab sampling work in the common top-k / top-k+top-p case.

  ## Sampling parameters

  The relevant sampling parameters are:

  - `top_k`:
    - integer cap on how many candidates are kept per row;
    - if active, only the highest-scoring `top_k` tokens remain eligible.
  - `top_p`:
    - nucleus threshold in cumulative probability space;
    - if active, tokens are kept until the cumulative probability reaches the threshold.
  - both can be used together:
    - first apply top-k to restrict the candidate set;
    - then apply top-p inside that surviving candidate set.

  In this proposal, the fast path targets:
  - top-k only
  - top-k + top-p

  ## Scope

  The proposed fast path applies when:

  - `k is not None`
  - active top-k is enabled for all rows, i.e. `k < vocab_size`
  - processed logits/logprobs are not requested

  I would not claim pure top-p-only support in this fast path, because exact pure top-p still needs identifying the nucleus over the full vocabulary unless a different exact algorithm is introduced.

  ## Correctness

  This optimization does not change the sampling mathematics.

  The baseline path computes:

  ```text
  softmax(mask_top_k_top_p(logits))
  ```

  where tokens outside the surviving set get zero probability.

  The candidate path computes the same distribution over the surviving candidates only:
  
  ```text
  softmax(logits[surviving_candidate_indices])
  ```

  and then gathers the sampled candidate back to the original vocabulary id.

  So this changes the computation layout, not the decoding semantics.

  I verified this by comparing the full-vocab masked probability tensor against the candidate-renormalized probability tensor:
  
  ```text
  top-k       distribution_match=True  max_abs_diff=2.980e-08
  top-k+top-p distribution_match=True  max_abs_diff=1.490e-08
  ```

  I also checked that sampled tokens always fall within the same surviving candidate set:

  ```text
  top-k       sampled_in_surviving_set=True
  top-k+top-p sampled_in_surviving_set=True
  ```

  The probability softmax remains fp32, matching the current native path. The existing random_sample implementation is reused, including the use_fp64_gumbel behavior.

  One note: sampled token ids are not guaranteed to be bitwise identical for the same global RNG seed, because the optimized path draws random noise only for surviving candidates instead of all
  vocabulary entries. The intended guarantee is distributional equivalence, not identical RNG consumption.

  ## Preliminary performance

  Tested on latest vLLM origin/main at commit 6f3da46, NVIDIA H100 PCIe, PyTorch 2.6.0+cu124.

  Baseline:

  top_k:  sampled from [1, 63]
  top_p: 1.0
  logits dtype: fp32

  latest-main apply_top_k_top_p
  (Triton when batch >= 8)
  + full-vocab softmax
  + random_sample

  Results:

  ``` text
   batch    vocab  baseline_ms candidate_ms  speedup
       1    32768        0.181        0.211     0.86x
       4    32768        0.204        0.203     1.01x
      16    32768        0.356        0.202     1.76x
      64    32768        0.359        0.201     1.79x
     256    32768        0.944        0.265     3.57x
    1024    32768        2.783        0.794     3.50x

       1   131072        0.208        0.202     1.03x
       4   131072        0.258        0.200     1.29x
      16   131072        1.217        0.201     6.04x
      64   131072        1.306        0.261     5.01x
     256   131072        4.469        0.785     5.69x
    1024   131072       14.617        2.277     6.42x
  ```
  The benefit is most visible for larger batch sizes and larger vocabularies. For very small batches the fast path can be neutral or slightly slower.

  ## Questions

  1. Does this candidate-set renormalization fast path make sense for the native sampler?
  2. Should the fast path be guarded by a batch-size or vocab-size threshold?
  3. Is torch.all(k < vocab_size) an acceptable guard, or should sampler metadata expose whether all requests have active top-k?
  4. Should this be implemented first as a small PyTorch-native fast path, or as a dedicated Triton kernel?

## Proposed Change

  Add a native candidate-set fast path for top-k / top-k+top-p sampling.

  The fast path would:
  1. select the top-k candidates for each row;
  2. optionally apply top-p within the surviving top-k set;
  3. renormalize only the surviving candidates;
  4. sample from that compact distribution;
  5. map the sampled candidate back to the original vocabulary id.

  This path should only be enabled when:
  - `k is not None`
  - active top-k is enabled for all rows
  - processed logits/logprobs are not requested

  The fallback path should remain unchanged for all other cases.

  I also added tests to verify:
  - the candidate-set probability distribution matches the masked full-vocab distribution;
  - sampled tokens always belong to the surviving candidate set.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
